### PR TITLE
APP-624 Support MessageML v2

### PR DESCRIPTION
**Pull request by robson-sanchez**
- Base parser class to generate MessageML v2 objects

### CODE REVIEW

**Comment by dnsymphony** [comment]
Nice docs!

Lot's of new classes here, but no matching unit tests.

Does Travis report unit test coverage?

**Comment by robson-sanchez** [comment]
@dnsymphony I'm not sure but I think Travis isn't reporting unit test coverage yet.

We need to check with Maurizio if it's required to configure some plugins on Maven to perform that report.

However, this PR has only the class MetadataParser with business logic. The other classes are only POJO to store the XML content. This class specifically has been tested.

**Review by dnsymphony — APPROVED** [review]

File: integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+   *
+   * Examples:
+   * - Replace '\n' to '<br/>' tags
+   * - Uppercase JSON field content
+   * - Scape characters
+   * - Include user ID and username retrieved from the User API
+   *
+   * @param input JSON input payload
+   */
+  protected abstract void preProcessInputData(JsonNode input);
```
Comment (dnsymphony) [inline integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java]: This doesn't need to be abstract.  You can provide an implementation that does nothing.  That way if subclass doesn't need to do any pre-processing, it doesn't need to implement a do-nothing method.

File: integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+   * This method must be implemented by the concrete parser classes to include additional fields
+   * on the output Entity JSON which weren't be processed directly from the JSON input payload.
+   *
+   * Example:
+   * - Array of JSON objects
+   *
+   * @param output Output Entity JSON
+   * @param input JSON input payload
+   */
+  protected abstract void postProcessOutputData(EntityObject output, JsonNode input);
```
Comment (dnsymphony) [inline integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java]: This doesn't need to be abstract. You can provide an implementation that does nothing. That way if subclass doesn't need to do any post-processing, it doesn't need to implement a do-nothing method.

File: integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    } catch (JsonProcessingException e) {
+      LOGGER.error("Fail to parse incoming payload", e);
+      return null;
+    }
+  }
+
+  /**
+   * Perform a pre-processing on the input data.
+   *
+   * This method must be implemented by the concrete parser classes to perform any data manipulation
```
Comment (dnsymphony) [inline integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java]: This method _can_ be overridden...

File: integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+      if (!entity.getContent().isEmpty()) {
+        root.addContent(object.getId(), entity);
+      }
+    }
+  }
+
+  /**
+   * Perform a post-processing on the output data.
+   *
+   * This method must be implemented by the concrete parser classes to include additional fields
```
Comment (dnsymphony) [inline integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java]: This method _can_ be overridden...

**Review by dnsymphony — CHANGES_REQUESTED** [review]

File: integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java (hunk 3 above)
Comment (robson-sanchez) [reply integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java]: DONE

File: integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java (hunk 1 above)
Comment (robson-sanchez) [reply integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java]: DONE

File: integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java (hunk 4 above)
Comment (robson-sanchez) [reply integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java]: DONE

File: integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java (hunk 2 above)
Comment (robson-sanchez) [reply integration-webhook/src/main/java/org/symphonyoss/integration/webhook/parser/metadata/MetadataParser.java]: DONE

**Comment by maoo** [comment]
@rsanchez-daitan @dnsymphony we can integrate with sonarqube in order to calculate testing coverage and other quality metrics; please read more on https://symphonyoss.atlassian.net/wiki/display/FM/Reporting#Reporting-SonarQube ; and - if interested - open an issue to request sonarqube to be enabled.

As an example, check the sonarqube dashboard of the symphony-java-client - https://sonarqube.com/dashboard?id=org.symphonyoss.symphony%3Asymphony-java-client

**Comment by robson-sanchez** [comment]
@dnsymphony please check again

**Review by dnsymphony — APPROVED** [review]

**Pull request merged**